### PR TITLE
Update Azure.Provisioning to latest

### LINKED ISSUE #6376: Update to GA version of Azure.Provisioning

### LINKED ISSUE #6474: New Azure SQL API has different naming conventions than Aspire 8
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

In Aspire 8, I had this code to provision an Azure SQL Server & Database:

```csharp
var sqlDatabase = builder.AddSqlServer("sql-server")
    .WithDataVolume()
    .PublishAsAzureSqlDatabase((_, _, _, sqlDatabases) =>
    {
        foreach (var db in sqlDatabases)
        {
            db.Properties.Sku = new SqlSku("S0") { Tier = "Standard" };
        }
    })
    .AddDatabase("evadexdb");
```

Upgrading to .NET 9.0-rc, and Aspire, I changed this code in accordance to the What's New docs:

```csharp
builder.Services.Configure<AzureResourceOptions>(options => options.ProvisioningContext.PropertyResolvers.Insert(0, new AspireV8ResourceNamePropertyResolver()));

var sqlDatabase = builder.AddAzureSqlServer("sql-server")
    .RunAsContainer(x => x.WithLifetime(ContainerLifetime.Persistent))
    .AddDatabase("evadexdb");
```

However, this has changed the naming on `sql-server` to `sql_server`, which is not a valid Azure SQL server name. `azd provision` returns this (version 1.10.3):

```
ERROR: deployment failed: error deploying infrastructure: deploying to subscription:

Deployment Error Details:
InvalidServerName: The server name 'sql_serverxr4o3alsj2qhw' cannot be empty or null. The server name can only be made up of lowercase letters 'a'-'z', the numbers 0-9 and the hyphen. The hyphen may not lead or trail in the server name. Please fix the server name and retry. Please contact Microsoft support if the issue persists.

TraceID: 489[42](https://github.com/PhantomSecurityGroup/EvadeX.Web/actions/runs/11490357271/job/31981252826#step:10:43)d2d159dc8f50fab7182f10d0124
Error: Process completed with exit code 1.
```

---

Taking a look at the different bison files generated:

**Aspire 8**
```js
resource sqlServer_XEivP4Ku5 'Microsoft.Sql/servers@2020-11-01-preview' = {
  name: toLower(take('sql-server${uniqueString(resourceGroup().id)}', 24))
  location: location
  tags: {
    'aspire-resource-name': 'sql-server'
  }
  properties: {
    version: '12.0'
    minimalTlsVersion: '1.2'
    publicNetworkAccess: 'Enabled'
    administrators: {
      administratorType: 'ActiveDirectory'
      login: principalName
      sid: principalId
      tenantId: subscription().tenantId
      azureADOnlyAuthentication: true
    }
  }
}

```

**Aspire 9 (w/ new Azure SQL API)**
```js
resource sql_server 'Microsoft.Sql/servers@2021-11-01' = {
  name: toLower(take('sql_server${uniqueString(resourceGroup().id)}', 24))
  location: location
  properties: {
    administrators: {
      administratorType: 'ActiveDirectory'
      login: principalName
      sid: principalId
      tenantId: subscription().tenantId
      azureADOnlyAuthentication: true
    }
    minimalTlsVersion: '1.2'
    publicNetworkAccess: 'Enabled'
    version: '12.0'
  }
  tags: {
    'aspire-resource-name': 'sql-server'
  }
}
```

### Expected Behavior

The previous name in Azure was `sql-serverxr4o3alsj2qhw` and should remain that, not `sql_serverxr4o3alsj2qhw`.

### Steps To Reproduce

_No response_

### Exceptions (if any)

_No response_

### .NET Version info

```
PS D:\source\PhantomSecurityGroup\EvadeX.Web> dotnet --info
.NET SDK:
 Version:           9.0.100-rc.2.24474.11
 Commit:            315e1305db
 Workload version:  9.0.100-manifests.0c3a5b37
 MSBuild version:   17.12.0-preview-24473-03+fea15fbd1

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.26120
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\9.0.100-rc.2.24474.11\

.NET workloads installed:
 [aspire]
   Installation Source: VS 17.11.35327.3
   Manifest Version:    8.2.1/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.aspire\8.2.1\WorkloadManifest.json
   Install Type:              Msi

Configured to use loose manifests when installing new manifests.

Host:
  Version:      9.0.0-rc.2.24473.5
  Architecture: x64
  Commit:       990ebf52fc

.NET SDKs installed:
  8.0.403 [C:\Program Files\dotnet\sdk]
  9.0.100-rc.2.24474.11 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 8.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.0-rc.2.24474.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.0-rc.2.24473.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 8.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 9.0.0-rc.2.24474.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  D:\source\PhantomSecurityGroup\EvadeX.Web\global.json

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

### Anything else?

Using the deprecated API resolves the issue.

### PR DESCRIPTION
**Pull request by eerhardt**
## Description

Respond to latest renames getting the API ready for GA release.

Fix https://github.com/dotnet/aspire/issues/6376

I'll use this PR until we get stable versions from @tg-msft.

Also fix an issue when using dashes `-` in resource names and using `AspireV8ResourceNamePropertyResolver`. We are prematurely converting the dash to an underscore, and the Azure resource name is not the same as it was in v8. This needs to be fixed together because the APIs we need to fix the issue are being made public in this new Azure.Provisioning version.

Fix #6474

## Checklist

- Is this feature complete?
  - [x] No. Follow-up changes expected. Still need stable versions plus more changes.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] Yes
    - Link to aspire-docs issue: https://github.com/dotnet/docs-aspire/issues/1652

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6390)

### CODE REVIEW

File: NuGet.config [hunk 1]
```diff
@@ -17,6 +17,7 @@
     <add key="dotnet9" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet9/nuget/v3/index.json" />
     <add key="dotnet-libraries" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-libraries/nuget/v3/index.json" />
     <add key="dotnet9-transport" value="https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet9-transport/nuget/v3/index.json" />
+    <add key="azure-sdk-devfeed" value="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json" />
```
Comment (eerhardt) [inline NuGet.config:20]: These NuGet.config changes won't be merged. We won't merge this PR until the `1.0.0` versions are on nuget.org.

File: src/Aspire.Hosting.Azure/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -17,18 +17,18 @@ Aspire.Hosting.Azure.AspireV8ResourceNamePropertyResolver
 Aspire.Hosting.Azure.AspireV8ResourceNamePropertyResolver.AspireV8ResourceNamePropertyResolver() -> void
 Aspire.Hosting.Azure.AzureProvisioningOptions
 Aspire.Hosting.Azure.AzureProvisioningOptions.AzureProvisioningOptions() -> void
-Aspire.Hosting.Azure.AzureProvisioningOptions.ProvisioningContext.get -> Azure.Provisioning.ProvisioningContext!
+Aspire.Hosting.Azure.AzureProvisioningOptions.ProvisioningBuildOptions.get -> Azure.Provisioning.ProvisioningBuildOptions!
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/PublicAPI.Unshipped.txt:20]: These API changes/renames are probably the most interesting changes in this PR.

File: src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs [hunk 3]
```diff
@@ -886,8 +885,7 @@ private void AddContainerRegistryParameters(ContainerAppConfiguration app)
     // This is a workaround to handle nested formattable strings until the bug is fixed.
     private static BicepValue<string> Interpolate(BicepValueFormattableString text)
     {
-        var formatStringBuilder = new StringBuilder();
-        var arguments = new List<BicepValue<string>>();
+        var bicepStringBuilder = new BicepStringBuilder();
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:888]: FYI - @davidfowl. We aren't able to get rid of our `Format` string parsing code from ReferenceExpression. But we are able to use the new `BicepStringBuilder` here.

**Comment by eerhardt** [comment]
Expecting 1 test failure due to https://github.com/Azure/azure-sdk-for-net/pull/46742

File: src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs (hunk 3 above)
Comment (eerhardt) [reply src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:888]: I think I could get rid of the BicepValueFormattableString class and just pass the `ReferenceExpression` around. No need in creating a new FormattableString object.

Comment (davidfowl) [reply src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:888]: We should be able to if you move this logic into the caller right?

Comment (eerhardt) [reply src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:888]: Yes, I can build up the BicepStringBuilder earlier and not have the BicepValueFormattableString at all.

Comment (davidfowl) [reply src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:885]: Ah yea, ok.

**Comment by eerhardt** [comment]
This is now ready for review. We have all the changes from the Azure.Provisioning libraries we need for 9.0.0 GA.

File: src/Aspire.Hosting.Azure.Sql/AzureSqlExtensions.cs [hunk 4]
```diff
@@ -256,9 +256,9 @@ private static void CreateSqlServer(
 
         foreach (var databaseNames in databases)
         {
-            var identifierName = Infrastructure.NormalizeIdentifierName(databaseNames.Key);
+            var BicepIdentifier = Infrastructure.NormalizeBicepIdentifier(databaseNames.Key);
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure.Sql/AzureSqlExtensions.cs:259]:
```suggestion
            var bicepIdentifier = Infrastructure.NormalizeBicepIdentifier(databaseNames.Key);
```

File: src/Aspire.Hosting.Azure.Sql/AzureSqlExtensions.cs [hunk 5]
```diff
@@ -256,9 +256,9 @@ private static void CreateSqlServer(
...
             var databaseName = databaseNames.Value;
-            var sqlDatabase = new SqlDatabase(identifierName)
+            var sqlDatabase = new SqlDatabase(BicepIdentifier)
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure.Sql/AzureSqlExtensions.cs:261]:
```suggestion
            var sqlDatabase = new SqlDatabase(bicepIdentifier)
```

**Comment by eerhardt** [comment]
This has all the changes from [[release/9.0] Update Azure.Provisioning to latest (dotnet/aspire#6489)](https://github.com/dotnet/aspire/pull/6489) merged into it. Looking for an approval so we can merge it into `main`.

**Review by davidfowl — APPROVED** [review]

**Pull request merged**